handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -237,36 +237,38 @@ protected void deallocate() {
         keyMaterialManager = context.keyMaterialManager();
         enableOcsp = context.enableOcsp;
 
-        Lock readerLock = context.ctxLock.readLock();
-        readerLock.lock();
-        try {
-            ssl = SSL.newSSL(context.ctx, !context.isClient());
-        } finally {
-            readerLock.unlock();
-        }
-        try {
-            networkBIO = SSL.bioNewByteBuffer(ssl, context.getBioNonApplicationBufferSize());
+        synchronized (this) {
+            Lock readerLock = context.ctxLock.readLock();
+            readerLock.lock();
+            try {
+                ssl = SSL.newSSL(context.ctx, !context.isClient());
```

Comment:
In response to @normanmaurer's https://github.com/netty/netty/pull/6955#discussion_r126482307 ... yes what I had in mind was something like:

```java
final int finalSsl;
Lock readerLock = context.ctxLock.readLock();
readerLock.lock();
try {
  finalSsl = SSL.newSSL(context.ctx, !context.isClient());
} finally {
  readerLock.unlock();
}

synchronized(this) {
  ssl = finalSsl;
...
}
```